### PR TITLE
Refactor latest anelyses

### PR DESCRIPTION
**Pull request by ChrOertlin**
### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: tests/cli/upload/test_cli_upload_nipt.py [hunk 1]
```diff
@@ -140,6 +142,24 @@ def test_nipt_statina_upload_auto(
...
+):
+    """Tests CLI command to upload without any analyses to upload"""
+
+    # GIVEN no analyses for upload
+    caplog.set_level(logging.DEBUG)
+
+    mocker.patch.object(NiptUploadAPI, "get_all_upload_analyses", return_value=None)
+
+    # WHEN uploading all NIPT cases
+    result = cli_runner.invoke(nipt_upload_all, [], obj=upload_context)
```
Comment (diitaz93) [inline tests/cli/upload/test_cli_upload_nipt.py:156]: use named arguments

File: tests/cli/upload/test_cli_upload_vogue.py [hunk 2]
```diff
@@ -86,3 +86,15 @@ def test_cli_upload_vogue_flowcells_no_days(upload_context: CGConfig, cli_runner
...
     assert result.exit_code != 0
+
+
+def test_cli_upload_vogue_bioinfo_all(upload_context: CGConfig, cli_runner: CliRunner):
+    """Testing cli for upload vogue bioinfo_all with correct argument"""
+
+    # GIVEN a vogue api
+
+    # WHEN running vogue upload bioinfo_all the days argument
+    result = cli_runner.invoke(bioinfo_all, ["-a", "2020-01-01"], obj=upload_context)
```
Comment (diitaz93) [inline tests/cli/upload/test_cli_upload_vogue.py:97]: use named arguments

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 3]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
+        assert analysis.family.internal_id in ["test_case_1", "yellowhog"]
+        assert not analysis.uploaded_to_vogue_at
+        assert analysis.completed_at < timestamp_in_2_weeks
+
+
+def test_get_analyses_uploaded_to_vogue_completed_after(
+    store_with_analyses_for_cases: Store,
+    timestamp_yesterday: datetime,
+):
+    """Test that an analysis can be fetched by case."""
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:67]: update docstring

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 4]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
+    for analysis in analyses:
+        assert analysis.family.internal_id in ["test_case_1", "yellowhog"]
+        assert not analysis.uploaded_to_vogue_at
+
+
+def test_get_analyses_uploaded_to_vogue_completed_before(
+    store_with_analyses_for_cases: Store,
+    timestamp_in_2_weeks: datetime,
+):
+    """Test that an analysis can be fetched by case."""
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:48]: update docstring

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 5]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
+    # GIVEN a database with an analysis and case
+
+    # WHEN fetching the latest analysis to upload to vogue completed after yesterday
+    analyses = store_with_analyses_for_cases.get_analysis_for_vogue_upload_completed_after(
+        completed_at_after=timestamp_yesterday
+    )
+
+    # THE the returned analysis was completed after yesterday
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:75]:
```suggestion
    # THEN the returned analysis was completed after yesterday
```

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 6]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
+        assert not analysis.uploaded_to_vogue_at
+        assert analysis.completed_at > timestamp_yesterday
+
+
 def test_get_latest_nipt_analysis_to_upload(
-    store_with_analyses_for_cases_not_uploaded_fluffy: Store, timestamp_now: datetime
+    store_with_analyses_for_cases_not_uploaded_fluffy: Store,
+    timestamp_now: datetime,
+    pipeline: str = Pipeline.FLUFFY,
 ):
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:86]: add docstring

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 7]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
     # GIVEN an analysis that is not delivery reported but there exists a newer analysis
 
-    # WHEN calling the analyses_to_delivery_report
-    analyses = (
-        store_with_analyses_for_cases_not_uploaded_fluffy.get_latest_nipt_analysis_to_upload()
+    # WHEN fetching the latest analysis to upload to nipt
+    analyses = store_with_analyses_for_cases_not_uploaded_fluffy.get_latest_analysis_to_upload_for_pipeline(
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:90]:
```suggestion
    analyses: List[Analysis] = store_with_analyses_for_cases_not_uploaded_fluffy.get_latest_analysis_to_upload_for_pipeline(
```

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 8]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
+    timestamp_now: datetime,
+    pipeline: str = Pipeline.MICROSALT,
 ):
     # GIVEN an analysis that is not delivery reported but there exists a newer analysis
 
-    # WHEN calling the analyses_to_delivery_report
-    analyses = (
-        store_with_analyses_for_cases_not_uploaded_microsalt.get_latest_microsalt_analysis_to_upload()
+    # WHEN fetching the latest analysis to upload to microsalt
+    analyses = store_with_analyses_for_cases_not_uploaded_microsalt.get_latest_analysis_to_upload_for_pipeline(
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:110]:
```suggestion
    analyses: List[Analysis] = store_with_analyses_for_cases_not_uploaded_microsalt.get_latest_analysis_to_upload_for_pipeline(
```

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 9]
```diff
@@ -26,37 +26,94 @@ def test_get_analysis_by_case(
...
-        assert analysis.pipeline == Pipeline.FLUFFY
+        assert analysis.pipeline == pipeline
 
 
 def test_get_latest_microsalt_analysis_to_upload(
-    store_with_analyses_for_cases_not_uploaded_microsalt: Store, timestamp_now: datetime
+    store_with_analyses_for_cases_not_uploaded_microsalt: Store,
+    timestamp_now: datetime,
+    pipeline: str = Pipeline.MICROSALT,
 ):
```
Comment (diitaz93) [inline tests/store/api/find/test_find_business_data_analysis.py:106]: add docstring

File: tests/store/api/status/test_analyses_to_clean.py [hunk 10]
```diff
@@ -1,71 +1,79 @@
...
 from cg.store import Store
 
 
 def test_analysis_included(analysis_store: Store, helpers):
-    """Tests that analyses that are uploaded are returned"""
+    """Tests that analyses that are uploaded are returned."""
 
     # GIVEN an analysis that is uploaded
-    timestamp = datetime.now()
+    timestamp_now = datetime.now()
```
Comment (diitaz93) [inline tests/store/api/status/test_analyses_to_clean.py:12]: this is a fixture already in `cg/test/conftest.py`.

File: tests/store/api/status/test_analyses_to_clean.py [hunk 11]
```diff
@@ -1,71 +1,79 @@
...
     assert analysis in analyses_to_clean
 
 
 def test_analysis_excluded(analysis_store: Store, helpers):
-    """Tests that analyses that are completed but lacks delivery report are returned"""
+    """Tests that analyses that are completed but lacks delivery report are returned."""
 
     # GIVEN an analysis that is not uploaded
-    timestamp = datetime.now()
+    timestamp_now = datetime.now()
```
Comment (diitaz93) [inline tests/store/api/status/test_analyses_to_clean.py:34]: same here

File: tests/store/api/status/test_analyses_to_clean.py [hunk 12]
```diff
@@ -1,71 +1,79 @@
...
     )
-    sample = helpers.add_sample(analysis_store, delivered_at=timestamp)
+    sample = helpers.add_sample(analysis_store, delivered_at=timestamp_now)
     analysis_store.relate_sample(family=analysis.family, sample=sample, status="unknown")
 
     # WHEN calling the analyses_to_clean
-    analyses_to_clean = analysis_store.analyses_to_clean().all()
+    analyses_to_clean = analysis_store.get_analyses_to_clean()
 
     # THEN this analyse should be returned
```
Comment (diitaz93) [inline tests/store/api/status/test_analyses_to_clean.py:45]:
```suggestion
    # THEN this analysis should be returned
```

File: tests/store/api/status/test_analyses_to_clean.py [hunk 13]
```diff
@@ -1,71 +1,79 @@
...
     assert analysis in analyses_to_clean
 
 
 def test_pipeline_excluded(analysis_store: Store, helpers):
-    """Tests that analyses are excluded depending on pipeline"""
+    """Tests that analyses are excluded depending on pipeline."""
 
     # GIVEN an analysis that is uploaded
     timestamp = datetime.now()
```
Comment (diitaz93) [inline tests/store/api/status/test_analyses_to_clean.py:79]: same here

**Review by diitaz93 — COMMENTED** [review]
Looks Great 💯 The only thing I think is important to change is the timestamp fixtures 😃

**Comment by ChrOertlin** [comment]
`cg upload vogue bioinfo-all -a 2040-01-01 --dry`
No analyses found for upload to trending.

` cg upload vogue bioinfo-all -b 2040-01-01 --dry`
Fetch latest version from bundle deepmole
Fetching files from bundle deepmole
Fetching files with tags in [multiqc-json]
Fetching files from version 255
continues

`cg upload vogue bioinfo-all -a 2000-01-01 --dry`
Fetch latest version from bundle readylion
Fetching files from bundle readylion
Fetching files with tags in [multiqc-json]
Fetching files from version 670
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/query.py:2577: SAWarning: Pathed join target File.version has already been joined to; skipping
  "been joined to; skipping" % prop
No multiqc file found in Housekeeper for case readylion
Fetch latest version from bundle lovingbug
Fetching files from bundle lovingbug
Fetching files with tags in [multiqc-json]
Fetching files from version 254

`cg upload nipt all --dry-run`
----------------- UPLOAD -----------------
*** NIPT UPLOAD ALL START ***
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating cgstats api
Set dry run to True

Uploading case: kinddragon
*** NIPT UPLOAD START ***
Set dry run to True
*** Statina UPLOAD START ***
Set dry run to True
Fetch latest version from bundle kinddragon
Fetching files with tags in [nipt,metrics]
Fetching files from version 82437
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/bin/cg", line 8, in <module>
    sys.exit(base())
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/upload/nipt/base.py", line 84, in nipt_upload_all
    context.invoke(nipt_upload_case, case_id=internal_id, dry_run=dry_run)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/upload/nipt/base.py", line 46, in nipt_upload_case
    context.invoke(batch, case_id=case_id, force=force, dry_run=dry_run)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/upload/nipt/statina.py", line 32, in batch
    statina_files: StatinaUploadFiles = nipt_upload_api.get_statina_files(case_id=case_id)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/upload/nipt/nipt.py", line 150, in get_statina_files
    case_id=case_id, tags=["nipt", "metrics"]
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/upload/nipt/nipt.py", line 79, in get_housekeeper_results_file
    f"No results files found in Housekeeper for NIPT case {case_id}."
cg.exc.HousekeeperFileMissingError: No results files found in Housekeeper for NIPT case kinddragon.

**Comment by ChrOertlin** [comment]
Nipt upload error is unrelated to changes

File: tests/cli/upload/test_cli_upload_nipt.py [hunk 14]
```diff
@@ -140,6 +145,46 @@ def test_nipt_statina_upload_auto(
...
+
+    # THEN the command should abort without raising an error
+    assert result.exit_code == 0
+    assert "No analyses found to upload" in caplog.text
+
+
+def test_nipt_statina_upload_auto_analysis_without_case(
+    upload_context: CGConfig, cli_runner: CliRunner, caplog, helpers, mocker
+):
+    """Tests CLI command to upload without any analyses to upload"""
```
Comment (Vince-janv) [inline tests/cli/upload/test_cli_upload_nipt.py:169]:
```suggestion
    """Tests CLI command to upload a non-existing case."""
```

File: tests/cli/upload/test_cli_upload_nipt.py [hunk 15]
```diff
@@ -140,6 +145,46 @@ def test_nipt_statina_upload_auto(
...
+
+    # WHEN uploading all NIPT cases
+    result = cli_runner.invoke(cli=nipt_upload_all, args=[], obj=upload_context)
+
+    # THEN the command should abort without raising an error
+    assert result.exit_code == 0
+    assert "No analyses found to upload" in caplog.text
+
+
+def test_nipt_statina_upload_auto_analysis_without_case(
```
Comment (Vince-janv) [inline tests/cli/upload/test_cli_upload_nipt.py:166]: I don't know if we need to test for this. It should be impossible

File: tests/cli/upload/test_cli_upload_nipt.py [hunk 16]
```diff
@@ -140,6 +145,46 @@ def test_nipt_statina_upload_auto(
     assert result.exit_code == 0
 
 
+def test_nipt_statina_upload_auto_without_analyses(
+    upload_context: CGConfig, cli_runner: CliRunner, caplog, helpers, mocker
+):
+    """Tests CLI command to upload without any analyses to upload"""
```
Comment (Vince-janv) [inline tests/cli/upload/test_cli_upload_nipt.py:151]:
```suggestion
    """Tests CLI command to upload without any analyses to upload."""
```

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 17]
```diff
@@ -26,37 +26,104 @@ def test_get_analysis_by_case(
...
+def test_get_analyses_uploaded_to_vogue(
+    store_with_analyses_for_cases: Store, case_id: str = "test_case_internal_id"
+):
+    """Test that an analysis can be fetched by case."""
+    # GIVEN a database with an analysis and case
+
+    # WHEN fetching the latest analysis to upload to vogue
+    analyses: List[Analysis] = store_with_analyses_for_cases.get_analyses_for_vogue_upload()
+
+    # THEN the analyses have not been uploaded to Vogue
```
Comment (Vince-janv) [inline tests/store/api/find/test_find_business_data_analysis.py:38]:
```suggestion
    # THEN the analyses should not have been uploaded to Vogue
```

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 18]
```diff
@@ -26,37 +26,104 @@ def test_get_analysis_by_case(
...
+    for analysis in analyses:
+        assert analysis.family.internal_id in ["test_case_1", "yellowhog"]
```
Comment (Vince-janv) [inline tests/store/api/find/test_find_business_data_analysis.py:40]: Nice test! Is this assert needed though?

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 19]
```diff
@@ -26,37 +26,104 @@ def test_get_analysis_by_case(
...
+    # WHEN fetching the latest analysis to upload to vogue that completed 2 weeks ago
+    analyses: List[
+        Analysis
+    ] = store_with_analyses_for_cases.get_analysis_for_vogue_upload_completed_before(
+        completed_at_before=timestamp_in_2_weeks
+    )
+
+    # THEN the returned analysis was completed earlier than 2 weeks ago
+    for analysis in analyses:
+        assert analysis.family.internal_id in ["test_case_1", "yellowhog"]
```
Comment (Vince-janv) [inline tests/store/api/find/test_find_business_data_analysis.py:60]: same here

File: tests/store/api/find/test_find_business_data_analysis.py [hunk 20]
```diff
@@ -26,37 +26,104 @@ def test_get_analysis_by_case(
...
+    # WHEN fetching the latest analysis to upload to nipt
+    analyses: List[
+        Analysis
+    ] = store_with_analyses_for_cases_not_uploaded_fluffy.get_latest_analysis_to_upload_for_pipeline(
+        pipeline=pipeline
     )
 
     # THEN only the newest analysis should be returned
     for analysis in analyses:
         assert analysis.family.internal_id in ["test_case_1", "yellowhog"]
```
Comment (Vince-janv) [inline tests/store/api/find/test_find_business_data_analysis.py:103]: and here

File: tests/store/api/test_base.py [hunk 21]
```diff
@@ -5,11 +5,49 @@
...
+    )
+    sample = helpers.add_sample(analysis_store, delivered_at=timestamp_now)
+    analysis_store.relate_sample(
+        family=analysis_oldest.family, sample=sample, status=PhenotypeStatus.UNKNOWN
+    )
+
+    # WHEN calling the analyses_to_delivery_report
+    analyses: Query = analysis_store._get_latest_analyses_for_cases_query()
+
+    # THEN analyses is a qeury
```
Comment (Vince-janv) [inline tests/store/api/test_base.py:48]:
```suggestion
    # THEN analyses is a query
```

File: tests/store/api/test_base.py [hunk 22]
```diff
@@ -5,11 +5,49 @@
...
     """Tests the _get_query function for all attributes of BaseHandler ie tables in the database."""
     assert isinstance(base_store._get_query(table=table), Query)
+
+
+def test_get_latest_analyses_for_cases_query(
+    analysis_store, helpers, timestamp_now, timestamp_yesterday
+):
+    """Tests that analyses that are not latest are not returned."""
+
+    # GIVEN an analysis that is not delivery reported but there exists a newer analysis
```
Comment (Vince-janv) [inline tests/store/api/test_base.py:23]: what does `is not delivery reported` mean, and why does it matter here?

File: tests/store/api/test_base.py [hunk 23]
```diff
@@ -5,11 +5,49 @@
...
+
+    # WHEN calling the analyses_to_delivery_report
+    analyses: Query = analysis_store._get_latest_analyses_for_cases_query()
+
+    # THEN analyses is a qeury
+    assert isinstance(analyses, Query)
+
+    # THEN only the newest analysis should be returned
+    assert analysis_newest in analyses
+    assert analysis_oldest not in analyses
```
Comment (Vince-janv) [inline tests/store/api/test_base.py:53]: Very good test though ⭐

File: tests/store/api/test_base.py (hunk 22 above)
Comment (ChrOertlin) [reply tests/store/api/test_base.py:23]:
```suggestion
    # GIVEN an analysis a newer analysis exists
```

**Pull request merged**